distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,969 @@
+from __future__ import annotations
+
+import asyncio
+import collections
+import logging
+import os
+import socket
+import struct
+import weakref
+from itertools import islice
+from typing import Any
+
+try:
+    import ssl
+except ImportError:
+    ssl = None  # type: ignore
+
+import dask
+
+from ..utils import ensure_ip, get_ip, get_ipv6
+from .addressing import parse_host_port, unparse_host_port
+from .core import Comm, CommClosedError, Connector, Listener
+from .registry import Backend
+from .utils import ensure_concrete_host, from_frames, to_frames
+
+logger = logging.getLogger(__name__)
+
+
+_COMM_CLOSED = object()
+
+
+def coalesce_buffers(
+    buffers: list[bytes],
+    target_buffer_size: int = 64 * 1024,
+    small_buffer_size: int = 2048,
+) -> list[bytes]:
+    """Given a list of buffers, coalesce them into a new list of buffers that
+    minimizes both copying and tiny writes.
+
+    Parameters
+    ----------
+    buffers : list of bytes_like
+    target_buffer_size : int, optional
+        The target intermediate buffer size from concatenating small buffers
+        together. Coalesced buffers will be no larger than approximately this size.
+    small_buffer_size : int, optional
+        Buffers <= this size are considered "small" and may be copied.
+    """
+    # Nothing to do
+    if len(buffers) == 1:
+        return buffers
+
+    out_buffers: list[bytes] = []
+    concat: list[bytes] = []  # A list of buffers to concatenate
+    csize = 0  # The total size of the concatenated buffers
+
+    def flush():
+        nonlocal csize
+        if concat:
+            if len(concat) == 1:
+                out_buffers.append(concat[0])
+            else:
+                out_buffers.append(b"".join(concat))
+            concat.clear()
+            csize = 0
+
+    for b in buffers:
+        size = len(b)
```

Comment:
Should we call `nbytes` here?

Reply:
No, if `b` is a memoryview here it's already cast to `'B'` format, so `len` is sufficient.